[#63482] Create Project Status Button Component

myabc opened the pull request:
⚠️ **~~This PR is based on #18478. Please review/merge first.~~**

# Ticket

https://community.openproject.org/wp/63482

# What are you trying to accomplish?

Implements a Project Status Button Component based on the existing OP Primer Status Component. Integrates the Status Button Component on the Project Settings > Information page.

In detail:

* Introduces `Projects::StatusButtonComponent` and its Lookbook Preview.
* Defines colours, icons as (Ruby) constants nested under `Project::Statuses` (c.f. `Meetings::Statuses`).
* Adds `Projects::StatusController` with `#update` action, routes and accompanying permissions.

## Screenshots

<img width="701" alt="Screenshot 2025-04-28 at 21 12 03" src="https://github.com/user-attachments/assets/114dd202-1315-4eb1-b280-5f009fd9337c" />
<img width="705" alt="Screenshot 2025-04-28 at 21 11 11" src="https://github.com/user-attachments/assets/a4e2d19f-57a7-4067-9057-6e6dd3ef81f0" />
<img width="700" alt="Screenshot 2025-04-28 at 21 11 02" src="https://github.com/user-attachments/assets/03b04953-2639-4f3a-b63a-41c7e729f04d" />
<img width="702" alt="Screenshot 2025-04-28 at 21 10 53" src="https://github.com/user-attachments/assets/68d13ebb-cf9a-4bc5-9ba9-11ece58f2c12" />

# What approach did you choose and why?

The implementation is based on the Meeting Status Button.

# Merge checklist

- [x] Added/updated tests
- [X] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

myabc committed:
[#63482] Implement Project Status Button Component

* Introduces `Projects::StatusButtonComponent` and its Lookbook Preview.
* Defines colours, icons as (Ruby) constants nested under
  `Project::Statuses` (c.f. `Meetings::Statuses`).
* Adds `Projects::StatusController` with `#update` action, routes and
  accompanying permissions.

myabc committed:
Implement have_turbo_stream RSpec matcher

Analgous to `assert_turbo_stream` provided by turbo-rails
`Turbo::TestAssertions`.

myabc committed:
Integrate Project Status Button in Settings form

myabc committed:
Clean up duplicate code in ProjectStatusHelper

myabc committed:
Update projects settings view spec expectations

myabc committed:
Make ShowComponent#initialize kwargs required

HDinger commented on config/locales/js-en.yml:
```diff
@@ -474,6 +474,7 @@ en:
     label_created_on: "created on"
     label_edit_comment: "Edit this comment"
     label_edit_status: "Edit the status of the work package"
+    label_edit_project_status: "Edit the status of the project"
```
This should be part of the normal `en.yml` file. The `js-en.yml` should ideally only contain strings that need to be accessed from Angular.

HDinger commented on app/views/highlighting/styles.css.erb:
```diff
@@ -9,6 +9,10 @@
 <% end %>
 
+<% Projects::Statuses::AVAILABLE.each do |project_status| %>
+<%= resource_color_css("project_status", project_status) %>
```
Missing indentation
```suggestion
  <%= resource_color_css("project_status", project_status) %>
```

myabc replied:
@HDinger I'll keep this as-is, as I think want to avoid extra indentation in the generated CSS.

HDinger commented on app/components/projects/status_button_component.rb:
```diff
@@ -0,0 +1,101 @@
+            title: I18n.t("js.label_edit_project_status"),
+            size: @size
+          },
+          menu_arguments: {
+            select_variant: :single
+          }
```
I think you can remove that and fallback to the default selection strategy which is `:none`. Then you can get rid of the additional CSS file and class you introduced.

HDinger commented on app/components/op_primer/status_button_component.sass:
```diff
@@ -0,0 +1,2 @@
+.op-status-button .ActionListItem-action--leading
```
I really don't like that, and I think it is not needed if you remove the `select_variant` from the menu_arguments of the Button (see my other comment).

HDinger reviewed (changes requested):
Works nicely 👍  I only have minor remarks.

myabc committed:
Fix test selector in ShowComponent spec

myabc committed:
Use DELETE method to unset project status

* Updates `StatusButtonComponent` to use `turbo-method: :delete` for
  "Not set" menu item.
* Adds `Projects::StatusController#destroy` action, routes and
  accompanying permissions.

myabc commented:
I've changed the base to `release/16.0` - this PR failed its first review before the release cut-off, but I am hoping we can still get this in to 16.0? I have addressed @HDinger's feedback.

In addition to making the changes requested by @HDinger, I have also improved/fixed error handling (the previous iteration of this PR ignored the unhappy path 🫨):

- ensuring the user is notified of both successful and failed updates via flash message.
- only rendering the Status Button with updated status if the update is successful.
- adding negative Controller specs.

<img width="1164" alt="image" src="https://github.com/user-attachments/assets/e7e44fe5-c358-418c-8e02-0a243300443c" />

<img width="905" alt="image" src="https://github.com/user-attachments/assets/30dcde90-afd9-4680-ac19-615459f1bfd1" />

myabc committed:
Remove outdated project status admin expectations

Edit settings feature already tests setting and unsetting project status
using the Project Status Button component.

The project status admin spec will likely be removed in a follow-up PR
(#18716 [OP Feature #63593](https://community.openproject.org/wp/63593))

myabc committed:
Add edit_project permission spec

myabc committed:
Add ProjectStatusHelper spec

myabc committed:
DRY up Projects::StatusController spec

Also aggregates failures where multiple expectations per example.

myabc commented on app/components/projects/status_button_component.rb:
```diff
@@ -0,0 +1,95 @@
+      href: project_status_path(project, status_code: status.value),
+      content_arguments: {
+        data: { turbo_method: status.value ? :put : :delete },
+        aria: { current: (true if status == @status) }
```
@bsatarnejad FYI. ActionMenu options use the `menuitem` role, so I think `aria-current` is appropriate here (vs. `aria-selected` or `aria-checked`) (we need to manually apply this ARIA attribute because we are using the [default `select_variant: :none`](https://primer.style/product/getting-started/rails/components/action_menu/#arguments))

myabc commented on app/components/projects/status_button_component.rb:
```diff
@@ -0,0 +1,95 @@
+      href: project_status_path(project, status_code: status.value),
+      content_arguments: {
+        data: { turbo_method: status.value ? :put : :delete },
+        aria: { current: (true if status == @status) }
```
since we are only interested in rendering `aria-current="true"`, we want this to evaluate to `true || nil`, not `true || false`.

myabc committed:
Document ProjectStatusHelper methods

HDinger reviewed (approved):
Lgtm 👍  I changed the base branch however back to back, since the feature freeze is long through.

HDinger commented:
mhh.. okay, I talked with QA about this again, and especially the part of the error handling is considered crutial to have in 16.0. In order, to avoid duplicate work, we will make an exception for this and merge it into 16.0
